core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala

```diff
@@ -0,0 +1,356 @@
+  // TODO: add assertions around every method
+
+  /**
+   * True if this block has been removed from the BlockManager and false otherwise.
+   * This field is used to communicate block deletion to blocked readers / writers (see its usage
+   * in [[BlockInfoManager]]).
+   */
+  var removed: Boolean = false
+
+  // TODO: Add timestamps on lock acquisitions
+}
+// In debugging mode, check that locks haven't been held for too long.
+// Every few minutes, dump debug info.
+
+/**
+ * Component of the [[BlockManager]] which tracks metadata for blocks and manages block locking.
+ *
+ * The locking interface exposed by this class is readers-writers lock. Every lock acquisition is
+ * automatically associated with a running task and locks are automatically released upon task
+ * completion or failure.
+ *
+ * This class is thread-safe.
+ */
+private[storage] class BlockInfoManager extends Logging {
+
+  private type TaskAttemptId = Long
+
+  /**
+   * Used to look up metadata for individual blocks. Entries are added to this map via an atomic
+   * set-if-not-exists operation ([[lockNewBlockForWriting()]]) and are removed
+   * by [[removeBlock()]].
+   */
+  @GuardedBy("this")
+  private[this] val infos = new mutable.HashMap[BlockId, BlockInfo]
+
+  /**
+   * Tracks the set of blocks that each task has locked for writing.
+   */
+  @GuardedBy("this")
+  private[this] val writeLocksByTask =
+    new mutable.HashMap[TaskAttemptId, mutable.Set[BlockId]]
+      with mutable.MultiMap[TaskAttemptId, BlockId]
+
+  /**
+   * Tracks the set of blocks that each task has locked for reading, along with the number of times
+   * that a block has been locked (since our read locks are re-entrant). This is thread-safe.
+   */
+  private[this] val readLocksByTask: LoadingCache[lang.Long, ConcurrentHashMultiset[BlockId]] = {
+    // We need to explicitly box as java.lang.Long to avoid a type mismatch error:
+    val loader = new CacheLoader[java.lang.Long, ConcurrentHashMultiset[BlockId]] {
+      override def load(t: java.lang.Long) = ConcurrentHashMultiset.create[BlockId]()
+    }
+    CacheBuilder.newBuilder().build(loader)
+  }
+
+  // ----------------------------------------------------------------------------------------------
+
+  /**
+   * Returns the current tasks's task attempt id (which uniquely identifies the task), or -1024
+   * if called outside of a task (-1024 was chosen because it's different than the -1 which is used
+   * in [[BlockInfo.writerTask]] to denote the absence of a write lock).
+   */
+  private def currentTaskAttemptId: TaskAttemptId = {
+    // TODO(josh): assert that this only happens on the driver?
+    // What about block transfer / getRemote()?
+    Option(TaskContext.get()).map(_.taskAttemptId()).getOrElse(-1024L)
+  }
+
+  /**
+   * Lock a block for reading and return its metadata.
+   *
+   * A single task can lock a block multiple times for reading, in which case each lock will need
+   * to be released separately.
```

Comment:
can you expand here on what happens if
- another task locked this for reading
- another task locked this for writing